server/data/data-api/src/test/java/org/apache/james/rrt/lib/CanSendFromContract.java

```diff
@@ -0,0 +1,108 @@
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.rrt.lib;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.apache.james.core.Domain;
+import org.apache.james.core.Username;
+import org.apache.james.rrt.api.CanSendFrom;
+import org.junit.jupiter.api.Test;
+
+public interface CanSendFromContract {
+
+    Domain DOMAIN = Domain.of("example.com");
+    Domain OTHER_DOMAIN = Domain.of("other.org");
+    Username USER = Username.of("user@example.com");
+    Username USER_ALIAS = Username.of("alias@example.com");
+    Username OTHER_USER = Username.of("other@example.com");
+
+    CanSendFrom canSendFrom();
+
+    void addDomain(Domain domain);
+    void addUser(Username username);
+
+    void addAliasMapping(Username alias, Username user) throws Exception;
+
+    void addDomainMapping(Domain alias, Domain domain) throws Exception;
+
+    void addGroupMapping(String group, Username user) throws Exception;
+
+
+    @Test
+    default void assertUserCanSendFromShouldBeFalseWhenSenderIsNotTheUser() {
+        addDomain(DOMAIN);
+        addUser(USER);
+        addUser(OTHER_USER);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, OTHER_USER)).isFalse();
+
+    }
+
+    @Test
+    default void assertUserCanSendFromShouldBeTrueWhenSenderIsTheUser() {
+        addDomain(DOMAIN);
+        addUser(USER);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, USER)).isTrue();
+    }
+
+    @Test
+    default void assertUserCanSendFromShouldBeFalseWhenSenderIsAnAliasOfAnotherUser() throws Exception {
+        addDomain(DOMAIN);
+        addUser(USER);
+        addAliasMapping(USER_ALIAS, OTHER_USER);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, USER_ALIAS)).isFalse();
+    }
+
+    @Test
+    default void assertUserCanSendFromShouldBeTrueWhenSenderIsAnAliasOfTheUser() throws Exception {
+        addDomain(DOMAIN);
+        addUser(USER);
+        addAliasMapping(USER_ALIAS, USER);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, USER_ALIAS)).isTrue();
+    }
+
+    @Test
+    default void assertUserCanSendFromShouldBeTrueWhenSenderIsAnAliasOfTheDomainUser() throws Exception {
+        addDomain(DOMAIN);
+        addDomain(OTHER_DOMAIN);
+        addUser(USER);
+        Username fromUser = Username.of(USER.getLocalPart() + "@" + OTHER_DOMAIN.asString());
+
+        addDomainMapping(OTHER_DOMAIN, DOMAIN);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, fromUser)).isTrue();
+    }
+
+    @Test
+    default void assertUserCanSendFromShouldBeFalseWhenWhenSenderIsAnAliasOfTheUserFromAGroupAlias() throws Exception {
+        addDomain(DOMAIN);
+        addUser(USER);
+
+        Username fromGroup = Username.of("group@example.com");
+
+        addGroupMapping("group@example.com", USER);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, fromGroup)).isFalse();
+
+    }
```

Comment:
what happens when the sender is an alias of an alias of the user?

Reply:
It should be handled, i have just added the corresponding test.